### PR TITLE
Quake test for proof robustness

### PR DESCRIPTION
**Pull request by mtzguido**
This introduces a `--quake` option whichs repeat SMT queries to test their robustness.

The basic idea is that setting `--quake N/M` will attempt each SMT query `M` times, and check that it succeeds at least `N` times (I sometimes `N` is called the low bar or threshold, and `M` the "high bar"). The repetition simulates exactly how F* discharges a VC, so each attempt will (for instance) retry with larger fuels on failure. As long as one "configuration" succeeds on each quake attempt, we call it valid.

Each attempt is run *in some unspecified different state*. Not necessarily obtained by varying the random seed (see ahead for why).

When the option is enabled (i.e. when the high bar is > 1), some summary info is printed on the command line:
```fstar
module D

open FStar.Mul

#set-options "--max_fuel 0 --max_ifuel 0"

val d (a b c : int) : Lemma ((a * b) * c = a * (b * c))
let d a b c = ()
```
```
$ ./bin/fstar.exe D.fst --quake 100
Quake: query (D.d, 1) succeeded 100/100 times
Verified module: D
All verification conditions discharged successfully
```

If we pass `--debug`, or use the interactive mode, we see some more lively output in `*Messages*` for each attempt:

If the low bar isn't met, F* will report all errors from all attempts, plus one stating that this is due to `--quake`. 
```fstar
module T

#set-options "--max_fuel 0 --max_ifuel 0"

let f () = assert False
```
```
$ ./bin/fstar.exe T.fst --quake 5
Quake: query (T.f, 1) succeeded 0/5 times
T.fst(5,11-5,23): (Error 325) Query (T.f, 1) failed the quake test, 0 out of 5 attempts succeded, but the threshold was 5
T.fst(5,18-5,23): (Error 19) assertion failed; The solver found a (partial) counterexample, try to spell your proof in more detail or increase fuel/ifuel
T.fst(5,18-5,23): (Error 19) assertion failed; The solver found a (partial) counterexample, try to spell your proof in more detail or increase fuel/ifuel
T.fst(5,18-5,23): (Error 19) assertion failed; The solver found a (partial) counterexample, try to spell your proof in more detail or increase fuel/ifuel
T.fst(5,18-5,23): (Error 19) assertion failed; The solver found a (partial) counterexample, try to spell your proof in more detail or increase fuel/ifuel
T.fst(5,18-5,23): (Error 19) assertion failed; The solver found a (partial) counterexample, try to spell your proof in more detail or increase fuel/ifuel
Verified module: T
6 errors were reported (see above)
```
I'm not so sure this a good choice though, since in emacs one might get a huge amount of errors, and the new one will (usually?) squiggly-line the entire function and hide the real problem. Maybe we should just not give a range to the new error? Or make it point to the `lid` being defined? 

- A question: Should we abort once the low bar is unreachable? I think we might want that for emacs, but not for a CI build which is meant to test robustness.

- About restarting, and Z3's random seed. It turns out we don't even need to change the seed to retry, since Z3 will have some randomly different behaviour after the pop. *But*, if we're using `--z3refresh`, then we do need to set it or we will re-run in the exact same state, so the code is currently special-casing for `--z3refresh`. If feels hacky though. What I originally wanted to do was to unconditionally set the seed via `(set-option smt.random_seed ..)` just before a `(check-sat)`, and I tried it, but turns out this seed change has absolutely no effect.

### CODE REVIEW

**Comment by jaybosamiya** [comment]
This is _awesome_!!! :smile: 

Few comments:
+ In the messages that are shown while quake is running, it would be awesome if all 3 of success+fail+remain are shown, rather than just success+remain. I know the number of failures can be calculated from the numbers on the screen, but they fly by fast enough that it'd be better to just show that, imho.
+ For the abort upon threshold failure, I agree that CI robustness build should _not_ abort early. However, for interactive use, early aborts are helpful since they reduce the amount of time the user needs to wait, right? Alternatively, if you introduce the failure counts in the messages, then the user themselves will `C-c C-c` kill it, right :upside_down_face:

**Comment by mtzguido** [comment]
Thanks Jay, I changed the message to something like this (also taking in @protz's suggestion regarding fuel):
```
...
Quake: so far query (D.assoc, 1) succeeded 23 times and failed 0 (2 runs remain)
Quake: so far query (D.assoc, 1) succeeded 24 times and failed 0 (1 runs remain)
Quake: query (D.assoc, 1) succeeded 25/25 times (best fuel=2, best ifuel=1)
```

**Comment by mtzguido** [comment]
The problem of interrupting F* is that the error will get lost, and you won't the see the source location highlighted. But if that's fixed, then it seems good enough to just `C-c C-c`!

**Comment by aseemr** [comment]
Thanks Guido, a high-level comment: should we combine duplicate errors to avoid seeing the same error 100 times. How does it happen right now, even with multiple queries with different fuel settings, we still see only one error?

File: src/smtencoding/FStar.SMTEncoding.Solver.fs [hunk 1]
```diff
@@ -651,34 +657,141 @@ let ask_and_report_errors env all_labels prefix query suffix =
...
+             * Why not always set it? Because it requires restarting the solver, which
+             * takes a long time.
+             *
+             * Why not use the (set-option smt.random_seed ..) command? Because
+             * it seems to have no effect just before a (check-sat), so it needs to be
+             * set early, which basically implies restarting.
+             *
+             * So we do this horrendous thing.
+             *)
+            if Options.z3_refresh ()
```
Comment (aseemr) [inline src/smtencoding/FStar.SMTEncoding.Solver.fs]: Agreed that looks a bit hacky. Another option might be to ignore `--z3_refresh` (with a console message saying that quaking does not restart z3) for the purposes of quaking.

**Comment by aseemr** [comment]
Added a couple of comments, looks good to me.

File: src/smtencoding/FStar.SMTEncoding.Solver.fs (hunk 1 above)
Comment (mtzguido) [reply src/smtencoding/FStar.SMTEncoding.Solver.fs]: I think I'd rather keep the hack than ignore --z3refresh. A funny thing I found is that for this:
```fstar
val lem (a b c : int) : Lemma ((a + b) * c = a * c + b * c)
let lem a b c = ()
```
I get 902/1000 successes normally, but 998/1000 with refresh. I can't really explain that discrepancy, but we might want to investigate these things in the future.

**Comment by mtzguido** [comment]
Good point on combining the errors, maybe we could even add a `(N times)` when we do. Currently they are just repeated, and too many of them are not very well displayed on emacs. Working on it now.

**Comment by mtzguido** [comment]
I merged it, comments on the code welcome anyway!

**Pull request merged**